### ISSUE TITLE
Playing pcap on 10G link, reports speeds faster than 10Gbps using netmap

### ISSUE THREAD

**Issue by morphyno** [appneta/tcpreplay#176]
Its reporting 12135.32Mbps (12Gbps), which is way above the 10G interface 

sudo /usr/bin/taskset -c 9 sudo /usr/bin/tcpreplay --netmap -i eth12 -tK /home/test.pcap
0.1422650785033.pcap
Switching network driver for eth12 to netmap bypass mode... done!
File Cache is enabled
Actual: 10943967 packets (4065006046 bytes) sent in 2.06 seconds.
Rated: 1516915559.6 Bps, 12135.32 Mbps, 4083898.92 pps
Flows: 879020 flows, 328018.97 fps, 10943967 flow packets, 0 non-flow
Statistics for network device: eth12
        Attempted packets:         10943967
        Successful packets:        10943967
        Failed packets:            0
        Truncated packets:         0
        Retried packets (ENOBUFS): 0
        Retried packets (EAGAIN):  0
Switching network driver for eth12 to normal mode... done!

What i'm noticing on the receiving side is that actually going about 5~6 Gbps (no flow control enabled).

So i'm suspecting there's a roundoff error somewhere in the calculation

**Comment by fklassen** [comment]
I think I have seen this with netmap. The netmap driver absorbs the packets, and has no mechanism for reporting when the packets have actually left the NIC. This only appears when using relatively small amounts of traffic on relatively high-capacity links. Can you try the same test with say `--loop 100` and send me a screenshot? I believe that a 2 second traffic burst is too difficult to quantify with tcprepaly. I find 10 seconds or greater will be more accurate.

Note that QuickTX has the opposite affect. It attempts to find out when the packets have left the NIC rather than when they were sent to the NIC. But with relatively small amounts of traffic, this results in reports of lower rates than reality. Until we get to the point where the NIC can report actual send times, we will have to live with a bit of an error.

**Comment by fklassen** [comment]
@morphyno I believe I have made improvements that will prevent this from happening. It may also fix other issues you have reported. I asked for advice from the author of netmap, and he gave me some tips that are reflected in PR #233. It will be available in v4.1.1 wish will be released shortly.

Thanks for the bug report. I'll keep this open for a while to see if you can confirm. It appears that the speed of your hardware may be a factor, so it would be nice to get feedback.

**Comment by fklassen** [comment]
I've been testing for several days, and it appears the PR #233 has fixed this. I'll close the issue, and please reopen if this reappears.

Will be part of 4.1.1 release.

**Comment by fklassen** [comment]
reopening. I did further optimizations and found I could reproduce the bug by sending 3000 packets with smallFlows.pcap. Larger and smaller bursts are OK.

```
root@portwell28:~# ./tcpreplay -i netmap:eth7 -Kt -L 3000 smallFlows.pcap 
Switching network driver for eth7 to netmap bypass mode... done!
File Cache is enabled
Actual: 3000 packets (2011013 bytes) sent in 0.001456 seconds.
Rated: 1381190200.0 Bps, 11049.52 Mbps, 2060439.56 pps
Flows: 1209 flows, 830357.14 fps, 14243 flow packets, 18 non-flow
Statistics for network device: eth7
	Attempted packets:         3061
	Successful packets:        3000
	Failed packets:            0
	Truncated packets:         0
	Retried packets (ENOBUFS): 0
	Retried packets (EAGAIN):  61
Switching network driver for eth7 to normal mode... done!
```

**Comment by fklassen** [comment]
fixed with PR #234

```
root@portwell28:~# ./tcpreplay -i netmap:eth7 -Kt -L 3000 smallFlows.pcap 
Switching network driver for eth7 to netmap bypass mode... done!
File Cache is enabled
Actual: 3000 packets (2011013 bytes) sent in 0.001796 seconds.
Rated: 1119717700.0 Bps, 8957.74 Mbps, 1670378.61 pps
Flows: 1209 flows, 673162.58 fps, 14243 flow packets, 18 non-flow
Statistics for network device: eth7
	Successful packets:        3000
	Failed packets:            0
	Truncated packets:         0
	Retried packets (ENOBUFS): 0
	Retried packets (EAGAIN):  58
Switching network driver for eth7 to normal mode... done!
root@portwell28:~# ./tcpreplay -i netmap:eth7 -Kt  smallFlows.pcap 
Switching network driver for eth7 to netmap bypass mode... done!
File Cache is enabled
Actual: 14261 packets (9216531 bytes) sent in 0.007483 seconds.
Rated: 1231662500.0 Bps, 9853.30 Mbps, 1905786.44 pps
Flows: 1209 flows, 161566.21 fps, 14243 flow packets, 18 non-flow
Statistics for network device: eth7
	Successful packets:        14261
	Failed packets:            0
	Truncated packets:         0
	Retried packets (ENOBUFS): 0
	Retried packets (EAGAIN):  398
Switching network driver for eth7 to normal mode... done!
```

**Comment by morphyno** [comment]
Thanks, this is excellent!
